metricbeat/module/vsphere/datastore_cluster/datastore_cluster.go

```diff
@@ -0,0 +1,102 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package datastore_cluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/view"
+	"github.com/vmware/govmomi/vim25/mo"
+
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/module/vsphere"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each network is defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+
+func init() {
+	mb.Registry.MustAddMetricSet("vsphere", "datastorecluster", New,
+		mb.WithHostParser(vsphere.HostParser),
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet type defines all fields of the MetricSet.
+type DatastoreClusterMetricSet struct {
+	*vsphere.MetricSet
+}
+
+// New creates a new instance of the MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	ms, err := vsphere.NewMetricSet(base)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create vSphere metricset: %w", err)
+	}
+	return &DatastoreClusterMetricSet{ms}, nil
+}
+
+func (m *DatastoreClusterMetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	client, err := govmomi.NewClient(ctx, m.HostURL, m.Insecure)
+	if err != nil {
+		return fmt.Errorf("error in NewClient: %w", err)
+	}
+
+	defer func() {
+		if err := client.Logout(ctx); err != nil {
+			m.Logger().Errorf("error trying to logout from vSphere: %w", err)
+		}
+	}()
+
+	c := client.Client
+
+	v, err := view.NewManager(c).CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"Network"}, true)
+	if err != nil {
+		return fmt.Errorf("error in CreateContainerView: %w", err)
+	}
+
+	defer func() {
+		if err := v.Destroy(ctx); err != nil {
+			m.Logger().Errorf("error trying to destroy view from vSphere: %w", err)
+		}
+	}()
+
+	var datastoreCluster []mo.StoragePod
+	err = v.Retrieve(ctx, []string{"StoragePod"}, []string{}, &datastoreCluster)
+	if err != nil {
+		return fmt.Errorf("error in Retrieve: %w", err)
+	}
+
+	for i := range datastoreCluster {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			reporter.Event(mb.Event{MetricSetFields: m.mapEvent(datastoreCluster[i])})
+		}
+	}
```

Comment:
Using select just for `ctx.Done()` isn't required and should ideally be used only when waiting for multiple channels. You could simply check the `ctx.Err()` instead like:
```suggestion
	for i := range datastoreCluster {
	    if ctx.Err() != nil {
	         return ctx.Err()
	    }
	           
	    reporter.Event(mb.Event{MetricSetFields: m.mapEvent(datastoreCluster[i])})
        }
```

Quoting the docs from the official [context](https://pkg.go.dev/context#Context) package:
>// If Done is closed, Err returns a non-nil error explaining why:
// Canceled if the context was canceled
// or DeadlineExceeded if the context's deadline passed.